src/NHibernate/Cache/SyncCacheLock.cs

```diff
@@ -6,34 +6,46 @@ namespace NHibernate.Cache
 {
 	class SyncCacheLock : ICacheLock
 	{
+		private readonly MonitorLock _monitorLock;
+
 		class MonitorLock : IDisposable
 		{
 			private readonly object _lockObj;
 
 			public MonitorLock(object lockObj)
 			{
-				Monitor.Enter(lockObj);
 				_lockObj = lockObj;
 			}
 
+			public IDisposable Lock()
+			{
+				Monitor.Enter(_lockObj);
+				return this;
+			}
+
 			public void Dispose()
 			{
 				Monitor.Exit(_lockObj);
 			}
 		}
 
+		public SyncCacheLock()
+		{
+			_monitorLock = new MonitorLock(this);
+		}
+
 		public void Dispose()
 		{
 		}
 
 		public IDisposable ReadLock()
 		{
-			return new MonitorLock(this);
+			return _monitorLock.Lock();
 		}
 
 		public IDisposable WriteLock()
 		{
-			return new MonitorLock(this);
+			return _monitorLock.Lock();
```

Comment:
But then it would be same lock for read and write. Is this expected?

Reply:
Yes. It's the same lock for read and write without this change. This locker doesn't distinguish read and write.

Reply:
What about this?


	class SyncCacheLock : ICacheLock
	{
		public void Dispose()
		{
			Monitor.Exit(this);
		}

		public IDisposable ReadLock()
		{
			return Lock();
		}

		public IDisposable WriteLock()
		{
			return Lock();
		}

                private IDisposable Lock()
		{
                        Monitor.Enter(this);
			return this;
		}

                //....

Reply:
No. ICacheLock is disposed separately from locks.